feat: add MSBuild rule to build MSBuild projects.

jsun-splunk opened the pull request:
This is an attempt to add a rule to build MSBuild project from bazel. This rule only supports `MSBuild.exe` from MSVC. `dotnet msbuild` is not supported.

[MSBuild, aka Microsoft Build Engine](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild?view=vs-2022) is a platform for building applications. This engine provides an XML schema for a project file that controls how the build platform processes and builds software. 

There are no source code to build from for MSBuild. There are also no prebuilt binaries that is available. Therefore, this rule is only availabe for pre-installed toolchain. This usually involves installing MSBuild along side MSVC.

MSBuild generates compile and link flags from project configuration and solution files. I have not been able to find a definitve way to override that so we can applied only the bazel generated flags. As a compromised, the generated bazel flags from cc_toolchain are added to a `msbuild.props` files and passed to MSBuild via the `-p:ForceImportAfterCppTargets` property. This at the minimum confirms bazel flags are applied.

matt-sm commented on .bazelci/config.yaml:
```diff
@@ -267,6 +267,8 @@ tasks:
       - "//..."
       # The min supported version of rules_rust is `>=7`
       - "-//rust/..."
+      # 6.5.0 seem to have a bug that does not honor target_compatible_with, so the msbuild target is getting triggered.
```
can you provide a link?

jsun-splunk replied:
I wasn't able to pinpoint an issue or PR. However, using a later version of bazel solves this problem.

The error was that the target under //msbuild_simple/... what had a target_compatible_with for windows was still being executed on linux.

matt-sm commented on foreign_cc/msbuild.bzl:
```diff
@@ -0,0 +1,199 @@
+"""# [MSBuild](#msbuild)
```
do you know if this will auto-magically get a dedicated doc page?

jsun-splunk replied:
That is what i'm hoping, but I haven't worked out a way to verify yet.

jsun-splunk commented on examples/msbuild_simple/BUILD.bazel:
```diff
@@ -0,0 +1,19 @@
+    out_lib_dir = "",
+    out_static_libs = ["mylib.lib"],
+    sln_file = "mysolution.sln",
+    target_compatible_with = ["@platforms//os:windows"],
```
I think this is true for MSVC msbuild, which is the one we adding. However, rule() doesn't allow you to set default `target_compatible_with`. It does support `exec_compatible_with` which also applies in this case. I've added that instead.

matt-sm replied:
It seems crazy to have to set this every time you use the rule.  Could we use a macro wrapper?

jsun-splunk replied:
I've done this locally via a wrapper, I may upstream it in a future MR.

jsun-splunk commented on foreign_cc/msbuild.bzl:
```diff
@@ -0,0 +1,199 @@
+"-p:PlatformToolset=", # e.g "v143" for Visual Studio 2022
+"-p:VCToolsRedistVersion=",
+"-p:VCInstallDir=<path to Visual C++ install dir>",
+"-p:VCInstallDir_170=<path to Visual C++ install dir for Visual Studios 2022",
```
fixed

jsun-splunk committed:
feat: Add rule for using MSBuild.exe from MSVC

jsun-splunk commented on foreign_cc/msbuild.bzl:
```diff
@@ -0,0 +1,297 @@
+        all_args.append("-t:{}".format(",".join(targets)))
+
+    expanded_args = expand_locations_and_make_variables(ctx, all_args, "args", data)
+    dep_paths = _msbuild_dep_paths(attrs.deps, inputs)
```
the dep paths that gets fed into the `msbuild.props` file now includes exact foreign cc type dep and bazel native type deps.

novas0x2a committed:
Merge branch 'main' into jsun-msbuild

novas0x2a committed:
msbuild: derive Configuration from compilation mode, add platform attr

Previously `configuration` was hardcoded to `Release`, ignoring Bazel's
compilation mode. It now defaults to `Debug` under `-c dbg` and `Release`
otherwise, mirroring the `cmake` rule.

This also adds a `platform` attribute. It is left unset by default so
MSBuild selects the platform declared in the solution/project file.
Platform is deliberately *not* derived from the target CPU: a `.sln` only
builds the `Configuration|Platform` pairs it declares, so forcing a
mismatched platform fails with `MSB4126`. Set it explicitly when the
project supports it (the example pins `x64`). `Platform` is now also
guarded as a reserved `properties` key alongside `Configuration`.

novas0x2a committed:
msbuild: move Windows exec constraint from rule to toolchain

The `msbuild` rule pinned `exec_compatible_with = [windows]` directly,
which is a hard constraint no custom toolchain can override. That
contradicts the rule's documented "bring your own MSBuild toolchain"
story — someone running MSBuild via a non-Windows exec platform (e.g. a
wine/remote setup) had no way to opt out.

Move the constraint onto `preinstalled_msbuild_toolchain` instead,
mirroring `preinstalled_nmake_toolchain`. Toolchain resolution now drives
the exec platform, so the default still requires Windows while a custom
toolchain can declare its own.

jsun-splunk committed:
Merge pull request #4 from bazel-contrib/msbuild-config-platform-exec

msbuild: derive Configuration from compilation mode and relax exec constraint
